pkg/kv/store.go

```diff
@@ -0,0 +1,97 @@
+package kv
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+)
+
+var (
+	ErrUnknownDriver   = errors.New("unknown driver")
+	ErrOperationFailed = errors.New("operation failed")
+	ErrMissingKey      = errors.New("missing key")
+	ErrMissingValue    = errors.New("missing value")
+	ErrNotFound        = errors.New("not found")
+	ErrClosedEntries   = errors.New("closed entries")
+)
+
+type Entry struct {
+	Key   []byte
+	Value []byte
+}
+
+type Driver interface {
+	// Open access to the database store. migrate kv if needed
+	Open(ctx context.Context, dsn string) (Store, error)
+
+	// Destroy will delete kv from storage
+	// Destroy(ctx context.Context, name string) error
+}
+
+type Store interface {
```

Comment:
This comment should probably arose sooner, but better late than never :)
How are planning to implement `Drop` or `DropByPrefix` we have for staging these days?

Reply:
1. Drop -> Delete
2. DropByPrefix -> DeleteByPrefix. Note that, not like staging - we don't have transaction, and it can be implemented by Scan + Delete with the current interface.

Reply:
We should probably add tests for concurrent iteration on one prefix and Delete on another prefix -- in all possible orders:

* ```
  iteration
               deletion
  ```
* ```
  iteration
        deletion
  ```
* ```
  iiiiiiiiiiteration
      deletion
  ```
* ```
    iteration
  deleeeeeeeetion
  ```
* ```
      iteration
  deletion
  ```
* ```
             iteration
  deletion
  ```